File: problems/p17/p17.cpp
/*
project euler problem 17
*/
#include <iostream>
#include <string.h>
using namespace std;

int solution(){
  int cache[1001] = {0};
  int sum = 0;
  cache[1] = strlen("one");
  cache[2] = strlen("two");
  cache[3] = strlen("three");
  cache[4] = strlen("four");
  cache[5] = strlen("five");
  cache[6] = strlen("six");
  cache[7] = strlen("seven");
  cache[8] = strlen("eight");
  cache[9] = strlen("nine");
  cache[10] = strlen("ten");
  cache[11] = strlen("eleven");
  cache[12] = strlen("twelve");
  cache[13] = strlen("thirteen");
  cache[14] = strlen("fourteen");
  cache[15] = strlen("fifteen");
  cache[16] = strlen("sixteen");
  cache[17] = strlen("seventeen");
  cache[18] = strlen("eighteen");
  cache[19] = strlen("nineteen");
  cache[20] = strlen("twenty");
  cache[30] = strlen("thirty");
  cache[40] = strlen("forty");
  cache[50] = strlen("fifty");
  cache[60] = strlen("sixty");
  cache[70] = strlen("seventy");
  cache[80] = strlen("eighty");
  cache[90] = strlen("ninety");

  for (int i = 21; i < 100; i++) {
    int tens_place = (i/10)*10;
    int ones_place = i - tens_place;
    cache[i] = cache[tens_place] + cache[ones_place];
  }

  for (int i = 100; i < 1000; i++) {
    int hundreds_place = i/100;
    int teens = i - (hundreds_place*100);
    if (teens == 0) {
      cache[i] = cache[hundreds_place] + strlen("hundred");
    }
    else{
      cache[i] = cache[hundreds_place] + strlen("hundredand") + cache[teens];
    }
  }

  cache[1000] = strlen("onethousand");

  for (size_t i = 1; i < 1001; i++) {
    sum += cache[i];
  }

  return sum;
}

int main(int argc, char const *argv[]) {
  cout << solution() << endl;
  return 0;
}

File: 17/p17.py
"""
project euler problem 17

i kept misspelling forty as fourty and so i kept trying a thousand different
ways to solve the problem and the real problem with my code was the spelling of the
word FORTY

i am So angry
"""

def please_god():
    cache = [0] * (1001)
    cache[0] = 0
    cache[1] = len('one')
    cache[2] = len('two')
    cache[3] = len('three')
    cache[4] = len('four')
    cache[5] = len('five')
    cache[6] = len('six')
    cache[7] = len('seven')
    cache[8] = len('eight')
    cache[9] = len('nine')
    cache[10] = len('ten')
    cache[11] = len('eleven')
    cache[12] = len('twelve')
    cache[13] = len('thirteen')
    cache[14] = len('fourteen')
    cache[15] = len('fifteen')
    cache[16] = len('sixteen')
    cache[17] = len('seventeen')
    cache[18] = len('eighteen')
    cache[19] = len('nineteen')
    cache[20] = len('twenty')
    cache[30] = len('thirty')
    cache[40] = len('forty')
    cache[50] = len('fifty')
    cache[60] = len('sixty')
    cache[70] = len('seventy')
    cache[80] = len('eighty')
    cache[90] = len('ninety')

    for i in range(21, 100):
        tens_place = (i//10)*10
        ones_place = i - tens_place
        cache[i] = cache[tens_place] + cache[ones_place]

    for i in range(100, 1000):
        hundreds_place = i//100
        teens = i - (hundreds_place*100)
        if teens == 0:
            cache[i] = cache[hundreds_place] + len('hundred')
        else:
            cache[i] = cache[hundreds_place] + len('hundredand') + cache[teens]

    cache[1000] = len('onethousand')

    print(sum(cache))

if __name__ == '__main__':
    please_god()
